proposals/NNNN-spirv-input-builtin.md

```diff
@@ -0,0 +1,159 @@
+# SPIR-V Input/Output built-ins
+
+ * Proposal: [NNNN](NNNN-spirv-input-builtin.md)
+ * Author(s): [Nathan Gauër](https://github.com/Keenuts)
+ * Status: **Design In Progress**
+
+## Introduction
+
+HLSL has semantic input/outputs parameters used to carry information
+to and from a shader, e.g: the system semantic `SV_GroupID` or `MY_SEMANTIC`,
+a user-defined semantic.
+
+In SPIR-V, those are translated to `Input` and `Output` variables,
+with either a `BuiltIn` or `Location` decoration.
+
+Example:
+
+```hlsl
+float4 vsmain(float4 Position : POSITION) : SV_POSITION
+{
+  return Position;
+}
+```
+
+Both `POSITION` and `SV_POSITION` will be represented as an `OpVariable`:
+
+- `POSITION` will have the `Input` storage class.
+- `SV_POSITION` will have the `Output` storage class.
+
+- `POSITION` will be decorated with `Location 0`.
+- `SV_POSITION` will be decorated with `BuiltIn Position`.
+
+In addition, users can define their own SPIR-V built-ins using inline SPIR-V:
+
+```hlsl
+[[vk::ext_builtin_input(/* NumWorkGroups */ 24)]]
+static const uint3 myBuiltIn;
+```
+
+This document explain how we plan on implementing those in Clang/LLVM.
+
+## Proposed Solution
+
+## Static behaving like an external & LLVM
+
+The current HLSL syntax for inline SPIR-V requires those built-ins to be
+`static [const]` variables. The rational was those variables were private to
+the invocation, and thus could be written/read without any concurrency issues.
+This was fine in DXC as we had our own AST->SPIR-V lowering.
+
+In Clang/LLVM, that's different: such static variable can be optimized in many
+ways:
+aggressive ways.
+ - A non-initialized `static const` is not allowed.
+ - Any dead-store to a `static int` can be removed.
+ - Meaning any pure computation parent to this store can be removed.
+
+Solving this can be done in 3 ways:
+ - Change the spec, allowing the syntax to be less "magical": use `extern`.
+ - Fixup the storage class of those variables in Clang.
+ - Emit an external constructor & destructor for those preventing optimization.
```

Comment:
Using a "constructor" for custom builtins is pretty similar to the approach we're using for handling semantics. We haven't yet worked on support for writing back to semantics, but I don't think that will be a problem.

In Clang today we generate two versions of each entry function. One that takes parameters, and one that doesn't. The version that takes no parameters is the true entry, and it initializes local allocas to the system values (builtins in SPIR-V parlance), then passes those as parameters into the secondary function which is called with always_inline.

This could be extended to emit global builtin initialization into the entry function as well.